### PR TITLE
package/base-files: caldata: use dd iflag fullblock

### PR DESCRIPTION
**Pull request by john-tho**
This dd flag ensures that the requested size
is retrieved from pipes or special filesystems (if available).

Without this flag, on multi-core systems,
Piped or special filesystem data can be truncated
when a size greater than PIPE_BUF is requested.

Fixes: FS#3494
Fixes: 7557e7f ("package/base-files: caldata: work around dd's
limitation")
Cc: Thibaut VARÈNE <hacks@slashdirt.org>

Signed-off-by: John Thomson <git@johnthomson.fastmail.com.au>

___

[Bug report: FS#3494](https://bugs.openwrt.org/index.php?do=details&task_id=3494)

http://lists.busybox.net/pipermail/busybox/2020-November/088326.html
> iflag=fullblock is the right solution. I added that feature a couple of
years ago. I thought it should be busybox's default behavior, but that goes
against the POSIX spec.
>
> POSIX dd specifies that one block == one read() call, unless modified by an
iflag. Read() calls can return partial data pretty often, especially when
reading from pipes. A filesystem read() will _usually_ give you as much
data as you request, but it's not guaranteed at all.
>
>So if you want to make _sure_ you're getting all of the data you want, the
only guaranteed safe approach is to use iflag=fullblock. It's good practice
to always use it with dd.

In the sysfs load:
I added the second `echo 1 > /sys/$DEVPATH/loading` to indicate to the driver there is a problem with the supplied firmware

https://www.kernel.org/doc/html/v5.4/driver-api/firmware/fallback-mechanisms.html#firmware-sysfs-loading-facility
> Using `echo 1 > /sys/$DEVPATH/loading`
> Will clean any previous partial load at once and make the firmware API return an error.

Tested on mikrotik hap ac2 & wap-g
I do not have an mtd or ubi caldata based device to test the other functions

@f00b4r0 Could you please take a look at this PR?

Somewhat related:

- We should be able to save a few cpu cycles by only running firmware hotplug caldata when $ACTION=add [not for remove as well](https://bugs.openwrt.org/index.php?do=details&task_id=3493)

### CODE REVIEW

**Comment by f00b4r0** [comment]
LGTM.
Personally I'd prefer to have $target as second argument and $offset last (i.e. if= of= count= skip= order, same as current), which also makes it possible to make $offset optional with a 0 default.
Other than that, ACK.
Thanks!

File: package/base-files/files/lib/functions/caldata.sh [hunk 1]
```diff
@@ -64,25 +76,30 @@ caldata_from_file() {
...
+	[ -d "$target_dir" ] || \
+		caldata_die "no sysfs dir to write: $target"
 
-	# can't fail now
 	echo 1 > /sys/$DEVPATH/loading
-	cat $source | dd of=/sys/$DEVPATH/data iflag=skip_bytes bs=$count skip=$offset count=1 2>/dev/null
-	echo 0 > /sys/$DEVPATH/loading
+	caldata_dd $source $offset $count $target
+	ret=$?
+	if [ ! $ret -eq 0 ]; then
```
Comment (jow-) [inline package/base-files/files/lib/functions/caldata.sh:97]: Minor nitpick - any reason to not simply do `[ $? != 0 ]` or even `if ! caldata_dd ...; then` ? We're not reusing `$ret` later and the syntax above is hard to understand (does it mean `(! $ret) == 0` or `!($ret == 0)`)

File: package/base-files/files/lib/functions/caldata.sh [hunk 2]
```diff
@@ -64,25 +76,30 @@ caldata_from_file() {
...
+	caldata_dd $source $offset $count $target || \
 		caldata_die "failed to extract calibration data from $source"
 }
 
 caldata_sysfsload_from_file() {
 	local source=$1
 	local offset=$(($2))
 	local count=$(($3))
+	local target="/sys/$DEVPATH/data"
+	local target_dir=$(dirname "$target")
```
Comment (jow-) [inline package/base-files/files/lib/functions/caldata.sh:88]: What about:

```
local target_dir="/sys/$DEVPATH"
local target="$target_dir/data"
```

Should save one subshell invocation

**Comment by john-tho** [comment]
Thank you @f00b4r0 and @jow-, I have implemented you suggestions.

File: package/base-files/files/lib/functions/caldata.sh (hunk 1 above)
Comment (john-tho) [reply package/base-files/files/lib/functions/caldata.sh]: Changed to `[  $? != 0 ]`, which make far more sense. Thank you!

File: package/base-files/files/lib/functions/caldata.sh (hunk 2 above)
Comment (john-tho) [reply package/base-files/files/lib/functions/caldata.sh]: Done, thank you.

**Pull request closed without merging**